### PR TITLE
Fix nilable gobj props

### PR DESCRIPTION
**Pull request by hugopl**
This fix nilable GObject properties, they were not going to the expected code path due to the fact they are presented as a Union type in macro world.

This also remove the set of fundamental_g_type functions.

### CODE REVIEW

File: src/bindings/g_object/object.cr [hunk 1]
```diff
@@ -134,17 +134,31 @@ module GObject
...
-                    {% elsif var.type < Enum %}
-                      {% if var.type.annotation(Flags) %}
-                        self.{{ var }} = raw.as(UInt32).unsafe_as({{ var.type }})
+                    {% if var.type.nilable? && var.type.union_types.size > 2 %}
+                      {% raise "Union types are not supported in GObject properties" %}
+                    {% end %}
+
+                    {% var_type = var.type.union_types.reject { |t| t == Nil }.first %}
+
+                    {% if var_type < GObject::Object %}
```
Comment (hugopl) [inline src/bindings/g_object/object.cr:143]: hmmm I think this should be `<=` instead o `<`.

**Comment by hugopl** [comment]
Ah, there's another issue this fixes, the Object `.cast`  call was inverted, the test wasn't catching it because the test didn't use a non-nilable GObject.

**Comment by BlobCodes** [comment]
GObject properties **must always** be nilable.
They can be set to nil using g_object_set.
This is the reason I defined `fundamental_g_type` - it does not allow defining non-nilable GObjects, gives a good compiler error and is very modular.

The other issues should be handled separately, this is not a good change.

**Comment by hugopl** [comment]
> GObject properties **must always** be nilable.
> They can be set to nil using g_object_set.

I don't think so. Yes be set to NULL by a `g_object_set` call from C code, but Crystal code can raise an exception in these cases, like the `.not_nil!` does at runtime.

IMO it's very useful to have a non-nilable GObject properties in Crystal code, to avoid writing null checks everywhere.

**Comment by hugopl** [comment]
Merging this, without you get a compiler error doing:

```
  @[GObject::Property]
  property hadjustment : Gtk::Adjustment?
```

**Pull request merged**